src/coreclr/jit/emitxarch.cpp

```diff
@@ -526,6 +526,16 @@ bool emitter::AreUpper32BitsZero(regNumber reg)
             }
         }
 
+        // This is a special case for cdq/cdqe/cwde.
+        // They always write to and sign-extend RAX.
```

Comment:
Will need to check `cdq` for EDX.

Reply:
`cmpxchg16b` also can write `RDX:RAX` (and reads `RCX:RBX`)

`div`, `idiv`, `imul`, and `mul` writes `RDX:RAX`

Might be others that I'm forgetting.

Reply:
I don't think we emit `cmpxchg16b` or even `cmpxchg8b` AFAIK. But I did need to check for `cmpxchg`.

Reply:
I am wondering if there is a good way to track such instructions or generalize the process with adequate asserts so we can catch the issues earlier? Tracking these issues with bad code gen are really hard to investigate.

Reply:
All I can think of is to make it more table-driven, otherwise a function that asks the question should be adequate.

Given we are only doing these checks here, we probably could create a function that simply asks "Did the instruction write to this register?" and the result is either 'true' or 'false', and if 'true' we can also provide if it zero-extended or not.
I think I'm going to have to do this anyway for https://github.com/dotnet/runtime/pull/82733 because I need to re-use the checks of whether or not a register was written to.

Reply:
I am more worried if we miss an entry of one of the instructions that we support, we might have similar bug. Is there a manual that we can refer and put all those instructions in the table or whatever data structure we have and also post the link of the manual so someone looking at the code in future can know where it is coming from?

Reply:
The Intel and AMD architecture manuals are the relevant resource.

The ones I listed are the ones that write two registers. There are others, like `cmpxchg`, which may write a single dedicated register.

It's likely just going to require someone going through the instructions and checking for cases that use a fixed/non-specified register (for input and/or output)